File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/zxtunes/RemoteCatalog.java
/**
 * @file
 * @brief Remote implementation of catalog
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.zxtunes;

import android.net.Uri;
import android.sax.Element;
import android.sax.RootElement;
import android.util.Xml;

import androidx.annotation.Nullable;

import org.xml.sax.SAXException;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Locale;

import app.zxtune.Log;
import app.zxtune.fs.HtmlUtils;
import app.zxtune.fs.http.MultisourceHttpProvider;

public final class RemoteCatalog implements Catalog {

  private static final String TAG = RemoteCatalog.class.getName();

  private static final String SITE = "http://zxtunes.com/";
  private static final String API = SITE + "xml.php?";
  private static final String ALL_AUTHORS_QUERY = API + "scope=authors&fields=nickname,name,tracks";
  //return nothing really, but required for more logical model
  private static final String ALL_TRACKS_QUERY = API
          + "scope=tracks&fields=filename,title,duration,date";
  private static final String AUTHOR_TRACKS_QUERY = ALL_TRACKS_QUERY + "&author_id=%d";
  private static final String DOWNLOAD_QUERY = SITE + "downloads.php?id=%d";

  private final MultisourceHttpProvider http;

  RemoteCatalog(MultisourceHttpProvider http) {
    this.http = http;
  }

  @Override
  public void queryAuthors(AuthorsVisitor visitor) throws IOException {
    Log.d(TAG, "queryAuthors()");
    final InputStream stream = http.getInputStream(Uri.parse(ALL_AUTHORS_QUERY));
    final RootElement root = createAuthorsParserRoot(visitor);
    performQuery(stream, root);
  }

  @Override
  public void queryAuthorTracks(Author author, TracksVisitor visitor) throws IOException {
    Log.d(TAG, "queryAuthorTracks(author=%d)", author.getId());
    queryTracks(visitor, String.format(Locale.US, AUTHOR_TRACKS_QUERY, author.getId()));
  }

  private void queryTracks(TracksVisitor visitor, String query) throws IOException {
    final InputStream stream = http.getInputStream(Uri.parse(query));
    final RootElement root = createModulesParserRoot(visitor);
    performQuery(stream, root);
  }

  @Override
  public boolean searchSupported() {
    return http.hasConnection();
  }

  @Override
  public void findTracks(String query, FoundTracksVisitor visitor) throws IOException {
    throw new IOException("Search is not supported on remote side");
  }

  public static Uri[] getTrackUris(int id) {
    final String query = String.format(Locale.US, DOWNLOAD_QUERY, id);
    return new Uri[]{Uri.parse(query)};
  }

  private void performQuery(InputStream httpStream, RootElement root)
          throws IOException {
    try {
      final InputStream stream = new BufferedInputStream(httpStream);
      Xml.parse(stream, Xml.Encoding.UTF_8, root.getContentHandler());
    } catch (SAXException e) {
      throw new IOException(e);
    } finally {
      httpStream.close();
    }
  }

  private static RootElement createAuthorsParserRoot(final AuthorsVisitor visitor) {
    final AuthorBuilder builder = new AuthorBuilder();
    final RootElement result = createRootElement();
    final Element list = result.getChild("authors");
    list.setStartElementListener(attributes -> {
      final Integer count = HtmlUtils.tryGetInteger(attributes.getValue("count"));
      if (count != null) {
        visitor.setCountHint(count);
      }
    });
    final Element item = list.getChild("author");
    item.setStartElementListener(attributes -> builder.setId(attributes.getValue("id")));
    item.setEndElementListener(() -> {
      final Author res = builder.captureResult();
      if (res != null) {
        visitor.accept(res);
      }
    });
    item.getChild("nickname").setEndTextElementListener(builder::setNickname);
    item.getChild("name").setEndTextElementListener(builder::setName);
    item.getChild("tracks").setEndTextElementListener(builder::setTracks);
    return result;
  }

  private static class AuthorBuilder {

    @Nullable
    private Integer id;
    @Nullable
    private String nickname;
    private String name = "";
    @Nullable
    private Integer tracks;

    AuthorBuilder() {
      reset();
    }

    final void setId(String val) {
      id = Integer.valueOf(val);
    }

    final void setNickname(String val) {
      nickname = val;
    }

    final void setName(String val) {
      name = val;
    }

    final void setTracks(String val) {
      tracks = Integer.valueOf(val);
    }

    @Nullable
    final Author captureResult() {
      final Author res = isValid() && hasTracks() ? new Author(id, nickname, name) : null;
      reset();
      return res;
    }

    private void reset() {
      id = null;
      tracks = null;
      nickname = null;
      name = "";
    }

    private boolean isValid() {
      return id != null && nickname != null;
    }

    private boolean hasTracks() {
      return tracks != null && tracks != 0;
    }
  }

  private static RootElement createModulesParserRoot(final TracksVisitor visitor) {
    final ModuleBuilder builder = new ModuleBuilder();
    final RootElement result = createRootElement();
    final Element list = result.getChild("tracks");
    list.setStartElementListener(attributes -> {
      final Integer count = HtmlUtils.tryGetInteger(attributes.getValue("count"));
      if (count != null) {
        visitor.setCountHint(count);
      }
    });
    final Element item = list.getChild("track");
    item.setStartElementListener(attributes -> builder.setId(attributes.getValue("id")));
    item.setEndElementListener(() -> {
      final Track result1 = builder.captureResult();
      if (result1 != null) {
        visitor.accept(result1);
      }
    });
    item.getChild("filename").setEndTextElementListener(builder::setFilename);
    item.getChild("title").setEndTextElementListener(builder::setTitle);
    item.getChild("duration").setEndTextElementListener(builder::setDuration);
    item.getChild("date").setEndTextElementListener(builder::setDate);
    return result;
  }

  private static class ModuleBuilder {

    @Nullable
    private Integer id;
    @Nullable
    private String filename;
    @Nullable
    private String title;
    @Nullable
    private Integer duration;
    @Nullable
    private Integer date;

    ModuleBuilder() {
      reset();
    }

    final void setId(String val) {
      id = Integer.valueOf(val);
    }

    final void setFilename(String val) {
      filename = val;
    }

    final void setTitle(String val) {
      title = val;
    }

    final void setDuration(String val) {
      duration = Integer.valueOf(val);
    }

    final void setDate(String val) {
      date = Integer.valueOf(val);
    }

    @Nullable
    final Track captureResult() {
      final Track res = isValid() ? new Track(id, filename, title, duration, date) : null;
      reset();
      return res;
    }

    private void reset() {
      id = null;
      duration = null;
      date = null;
      filename = null;
      title = "";
    }

    private boolean isValid() {
      return id != null && filename != null;
    }
  }

  private static RootElement createRootElement() {
    //TODO: check root tag version
    return new RootElement("zxtunes");
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/zxtunes/RemoteCatalog.kt
/**
 * @file
 * @brief Remote implementation of catalog
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs.zxtunes

import android.net.Uri
import android.sax.RootElement
import app.zxtune.Logger
import app.zxtune.fs.HtmlUtils.tryGetInteger
import app.zxtune.fs.http.MultisourceHttpProvider
import app.zxtune.utils.Xml
import app.zxtune.utils.ifNotNulls
import org.xml.sax.Attributes
import java.io.BufferedInputStream
import java.io.IOException

private val LOG = Logger(RemoteCatalog::class.java.name)

class RemoteCatalog internal constructor(private val http: MultisourceHttpProvider) : Catalog {
    override fun queryAuthors(visitor: Catalog.Visitor<Author>) {
        LOG.d { "queryAuthors()" }
        val root = createAuthorsParserRoot(visitor)
        performQuery(Uris.forAllAuthors(), root)
    }

    override fun queryAuthor(id: Int): Author? {
        LOG.d { "queryAuthor(id=${id})" }
        val visitor = object : Catalog.Visitor<Author> {
            var result: Author? = null
            override fun accept(obj: Author) {
                require(result == null)
                result = obj
            }
        }
        val root = createAuthorsParserRoot(visitor)
        performQuery(Uris.forAuthor(id), root)
        return visitor.result
    }

    override fun queryAuthorTracks(author: Author, visitor: Catalog.Visitor<Track>) {
        LOG.d { "queryAuthorTracks(author=${author.id})" }
        queryTracks(visitor, Uris.forTracks(author))
    }

    private fun queryTracks(visitor: Catalog.Visitor<Track>, uri: Uri) {
        val root = createModulesParserRoot(visitor)
        performQuery(uri, root)
    }

    override fun searchSupported() = http.hasConnection()

    override fun findTracks(query: String, visitor: Catalog.FoundTracksVisitor) =
        throw IOException("Search is not supported on remote side")

    private fun performQuery(uri: Uri, root: RootElement) =
        http.getInputStream(uri).let { httpStream ->
            Xml.parse(BufferedInputStream(httpStream), root.contentHandler)
        }

    companion object {
        @JvmStatic
        fun getTrackUris(id: Int) = arrayOf(Uris.forDownload(id))

        @JvmStatic
        fun getImageUris(id: Int) = arrayOf(Uris.forPhoto(id))
    }
}

private class Uris {
    private val delegate = Uri.Builder()

    init {
        delegate.scheme("http") //TODO: https
            .authority("zxtunes.com")
    }

    fun forApi(): Uri = delegate.appendPath("xml.php").build()

    fun forDownloads(): Uri = delegate.appendPath("downloads.php").build()

    fun forPhoto(id: Int): Uri = delegate.appendPath("photo").appendEncodedPath("${id}.jpg").build()

    fun scope(scope: String) = apply {
        delegate.appendQueryParameter("scope", scope)
    }

    fun fields(fields: String) = apply {
        delegate.appendQueryParameter("fields", fields)
    }

    fun entity(entry: String, id: Int) = apply {
        delegate.appendQueryParameter(entry, id.toString())
    }

    companion object {
        private fun forAuthors() = Uris().scope("authors").fields("nickname,name,tracks,photo")

        fun forAllAuthors() = forAuthors().forApi()

        fun forAuthor(id: Int) = forAuthors().entity("id", id).forApi()

        fun forTracks(author: Author) =
            Uris().scope("tracks").fields("filename,title,duration,date")
                .entity("author_id", author.id).forApi()

        fun forDownload(id: Int) = Uris().entity("id", id).forDownloads()

        fun forPhoto(id: Int) = Uris().forPhoto(id)
    }
}

private class AuthorBuilder {
    private var id: Int? = null
    private var nickname: String? = null
    private var name = ""
    private var tracks: Int = 0
    private var hasPhoto = false

    init {
        reset()
    }

    fun setId(value: String?) {
        id = tryGetInteger(value)
    }

    fun setNickname(value: String?) {
        nickname = value
    }

    fun setName(value: String?) {
        name = value.orEmpty()
    }

    fun setTracks(value: String?) {
        tracks = tryGetInteger(value) ?: 0
    }

    fun setHasPhoto(value: String?) {
        hasPhoto = 1 == tryGetInteger(value)
    }

    fun captureResult() =
        ifNotNulls(id, nickname, name.takeIf { tracks > 0 }, hasPhoto, ::Author).also {
            reset()
        }

    private fun reset() {
        id = null
        nickname = null
        tracks = 0
        name = ""
        hasPhoto = false
    }
}

private class ModuleBuilder {
    private var id: Int? = null
    private var filename: String? = null
    private var title: String = ""
    private var duration: Int? = null
    private var date: Int? = null

    init {
        reset()
    }

    fun setId(value: String?) {
        id = tryGetInteger(value)
    }

    fun setFilename(value: String?) {
        filename = value
    }

    fun setTitle(value: String?) {
        title = value.orEmpty()
    }

    fun setDuration(value: String?) {
        duration = tryGetInteger(value)
    }

    fun setDate(value: String?) {
        date = tryGetInteger(value)
    }

    fun captureResult() = ifNotNulls(id, filename) { id, filename ->
        Track(id, filename, title, duration, date)
    }.also {
        reset()
    }

    private fun reset() {
        id = null
        duration = null
        date = null
        filename = null
        title = ""
    }
}

private fun createAuthorsParserRoot(visitor: Catalog.Visitor<Author>) = createRootElement().apply {
    getChild("authors").run {
        val builder = AuthorBuilder()
        getChild("author").run {
            setStartElementListener { attributes: Attributes ->
                builder.setId(attributes.getValue("id"))
            }
            setEndElementListener {
                builder.captureResult()?.let {
                    visitor.accept(it)
                }
            }
            getChild("nickname").setEndTextElementListener(builder::setNickname)
            getChild("name").setEndTextElementListener(builder::setName)
            getChild("tracks").setEndTextElementListener(builder::setTracks)
            getChild("photo").setEndTextElementListener(builder::setHasPhoto)
        }
    }
}

private fun createModulesParserRoot(visitor: Catalog.Visitor<Track>) = createRootElement().apply {
    val builder = ModuleBuilder()
    getChild("tracks").getChild("track").run {
        setStartElementListener { attributes: Attributes ->
            builder.setId(attributes.getValue("id"))
        }
        setEndElementListener {
            builder.captureResult()?.let {
                visitor.accept(it)
            }
        }
        getChild("filename").setEndTextElementListener(builder::setFilename)
        getChild("title").setEndTextElementListener(builder::setTitle)
        getChild("duration").setEndTextElementListener(builder::setDuration)
        getChild("date").setEndTextElementListener(builder::setDate)
    }
}

//TODO: check root tag version
private fun createRootElement() = RootElement("zxtunes")
